File: Nostos.Shared/Dtos/ImportDtos.cs
using System;
using System.Collections.Generic;

namespace Nostos.Shared.Dtos;

// --- BACKGROUND IMPORT PROGRESS (issue #189) ---
// A dedicated feed for work in flight, deliberately NOT the library query.
// The library answers "what do I own, sorted and filtered how I asked"; this
// answers "what is being imported right now", which must be visible no matter
// how the library is currently sorted or filtered — an import is not a search
// result and must not disappear because the user filtered to Favourites.
//
// Its own vocabulary rather than a reuse of ProviderAcquisitionDto, because an
// entry here can be either a live job OR a book row whose import a server
// restart interrupted (see the reconciliation worker). The former has no book
// yet in the general case; the latter has no job at all.

/// <summary>
/// One row of the "Imports in Progress" surface. Identified by <see cref="Id"/>,
/// which is the acquisition job id for a live job and the book id for an entry
/// recovered from a previous process's interrupted import.
/// </summary>
public sealed record ImportActivityDto(
    /// <summary>Stable id for this entry: the job id, or the book id when reconciled.</summary>
    string Id,
    /// <summary>"job" (a job in this process) or "reconciled" (an interrupted import).</summary>
    string Source,
    /// <summary>queued | running | succeeded | failed | cancelled</summary>
    string State,
    /// <summary>resolving | downloading | transcoding | importing | done | failed …</summary>
    string Stage,
    /// <summary>
    /// 0-100, already capped at 99 unless the state is terminal Succeeded. The
    /// cap is applied server-side so no client can read "100%" and offer a file
    /// that has not landed yet.
    /// </summary>
    int Percent,
    string? Detail,
    string? ProviderId,
    string? ExternalId,
    string? AssetId,
    Guid? BookId,
    /// <summary>The book row's title, from the moment the user confirmed the import.</summary>
    string? Title,
    string? Author,
    /// <summary>Nostos-relative cover URL for the book row, or null when it has none yet.</summary>
    string? CoverUrl,
    string? ErrorCode,
    string? Message,
    DateTime CreatedAt,
    DateTime UpdatedAt,
    /// <summary>False when the source cannot be re-imported from this entry (no provider/external id).</summary>
    bool CanRetry);

File: Nostos.Frontend/src/app/core/dtos/import.dtos.ts
import { AcquisitionState } from './provider.dtos';

/**
 * One row of the "Imports in Progress" surface.
 *
 * Deliberately NOT a library book: it describes work in flight, which must stay
 * visible whatever the library is filtered or sorted by. Two kinds of entry
 * share the shape — a live job in this server process (`source: 'job'`), and a
 * book row whose import a server restart interrupted (`source: 'reconciled'`,
 * which arrives already terminal).
 */
export interface ImportActivity {
  /** Stable id: the job id, or the book id when the entry is a reconciled one. */
  id: string;
  source: 'job' | 'reconciled';
  state: AcquisitionState;
  stage: string;
  /**
   * 0-100, already capped at 99 by the server unless the state is Succeeded —
   * the client must never render 100% for an import that has not landed.
   */
  percent: number;
  detail: string | null;
  providerId: string | null;
  externalId: string | null;
  assetId: string | null;
  bookId: string | null;
  /** The book row's title, present from the moment the user confirmed. */
  title: string | null;
  author: string | null;
  /** Nostos-relative cover URL for the book row, or null while it has none. */
  coverUrl: string | null;
  errorCode: string | null;
  message: string | null;
  createdAt: string;
  updatedAt: string;
  /** False when this entry cannot be re-imported (no provider/external id). */
  canRetry: boolean;
}

/** States after which nothing about an entry will change again. */
export const IMPORT_TERMINAL_STATES: ReadonlySet<AcquisitionState> = new Set([
  'succeeded',
  'failed',
  'cancelled',
]);

/** States that are still doing work. */
export const IMPORT_IN_FLIGHT_STATES: ReadonlySet<AcquisitionState> = new Set([
  'queued',
  'running',
]);

export function isImportTerminal(activity: ImportActivity): boolean {
  return IMPORT_TERMINAL_STATES.has(activity.state);
}

export function isImportInFlight(activity: ImportActivity): boolean {
  return IMPORT_IN_FLIGHT_STATES.has(activity.state);
}

/**
 * The short badge text for an entry.
 *
 * The server's stages are deliberately few and provider-agnostic, and several of
 * them are the same thing to a reader: resolving a manifest, checking free space
 * and fetching bytes are all "downloading" from where they sit. Two labels the
 * user actually cares about — Downloading and Transcoding — plus the terminal
 * ones.
 */
export function importStageLabel(activity: ImportActivity): string {
  switch (activity.state) {
    case 'succeeded':
      return 'Done';
    case 'failed':
      return 'Failed';
    case 'cancelled':
      return 'Cancelled';
    default:
      break;
  }

  switch (activity.stage) {
    case 'transcoding':
    case 'assembling':
    case 'importing':
      return 'Transcoding';
    case 'queued':
    case 'starting':
      return 'Queued';
    case 'done':
      return 'Done';
    default:
      return 'Downloading';
  }
}
